server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceFamily.java

```diff
@@ -0,0 +1,34 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.search.aggregations.support;
+
+import org.elasticsearch.index.query.QueryShardContext;
+import org.elasticsearch.script.AggregationScript;
+import org.elasticsearch.search.DocValueFormat;
+
+public interface ValuesSourceFamily {
+    ValuesSource getEmpty();
+
+    ValuesSource getScript(AggregationScript.LeafFactory script, ValueType scriptValueType);
+
+    ValuesSource getField(FieldContext fieldContext, AggregationScript.LeafFactory script);
+
+    ValuesSource replaceMissing(ValuesSource valuesSource, Object rawMissing, DocValueFormat format, QueryShardContext queryShardContext);
```

Comment:
I think we only need the QueryShardContext for the value of `now`? If that's the case, I'd rather like that we replace this QueryShardContext with `LongProducer now` which would make this class easier to unit test?

Reply:
Yeah, good call, that'll be a lot cleaner.